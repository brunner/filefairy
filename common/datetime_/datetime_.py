#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import datetime
import pytz
import re

_est = pytz.timezone('America/New_York')
_pst = pytz.timezone('America/Los_Angeles')


def datetime_as_pst(d):
    return d.astimezone(_pst)


def datetime_datetime_est(*args):
    return _est.localize(datetime.datetime(*args))


def datetime_datetime_pst(*args):
    return _pst.localize(datetime.datetime(*args))


def datetime_now():
    return _pst.localize(datetime.datetime.now())


def decode_datetime(s):
    return datetime_datetime_pst(*map(int, re.findall('\d+', s[:-6])))


def encode_datetime(d):
    return datetime_as_pst(d).isoformat()


def suffix(day):
    return 'th' if 11 <= day <= 13 else {
        1: 'st',
        2: 'nd',
        3: 'rd'
    }.get(day % 10, 'th')


def timedelta(then, now):
    diff = now - then
    s, d = diff.seconds, diff.days

    if s < 0 or d < 0:
        return timedelta(now, then)

    if d > 0:
        s += 86400 * d

    h, m = s // 3600, (s // 60) % 60
    hm = []
    if h > 0:
        hm.append('{}h'.format(h))
    hm.append('{}m'.format(m))

    return ' '.join(hm)


def timestamp(date):
    return date.strftime('%H:%M:%S %Z (%Y-%m-%d)')
