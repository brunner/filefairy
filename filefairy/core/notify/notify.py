#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from enum import Enum


class Notify(Enum):
    BASE = 1

    EXPORTS_EMAILS = 2

    FAIRYLAB_DAY = 3

    FAIRYLAB_DEPLOY = 4

    LEAGUEFILE_DOWNLOAD = 5

    LEAGUEFILE_FINISH = 6

    LEAGUEFILE_START = 7

    LEAGUEFILE_YEAR = 8

    STATSPLUS_SIM = 9

    OTHER = 10
