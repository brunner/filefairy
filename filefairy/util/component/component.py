#!/usr/bin/env python3
# -*- coding: utf-8 -*-


def anchor(url, content):
    return '<a href="{}">{}</a>'.format(url, content)


def card(href='', title='', info='', table=None, ts='', success='', danger=''):
    return {
        'href': href,
        'title': title,
        'info': info,
        'table': table,
        'ts': ts,
        'success': success,
        'danger': danger
    }


def span(classes, text):
    return '<span class="{}">{}</span>'.format(' '.join(classes), text)


def table(clazz='border mt-3', hcols=None, bcols=None, head=None, body=None):
    return {
        'clazz': clazz,
        'hcols': hcols,
        'bcols': bcols,
        'head': head,
        'body': body
    }
