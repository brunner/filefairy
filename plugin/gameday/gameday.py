#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import copy
import json
import os
import re
import sys
from functools import partial

_path = os.path.dirname(os.path.abspath(__file__))
_root = re.sub(r'/plugin/gameday', '', _path)
sys.path.append(_root)
from api.registrable.registrable import Registrable  # noqa
from core.notify.notify import Notify  # noqa
from core.response.response import Response  # noqa
from util.component.component import anchor  # noqa
from util.component.component import bold  # noqa
from util.component.component import secondary  # noqa
from util.component.component import table  # noqa
from util.datetime_.datetime_ import decode_datetime  # noqa
from util.file_.file_ import recreate  # noqa
from util.statslab.statslab import parse_player  # noqa
from util.team.team import divisions  # noqa
from util.team.team import encoding_to_decoding  # noqa
from util.team.team import encoding_to_nickname  # noqa
from util.team.team import encoding_to_teamid  # noqa
from util.team.team import logo_absolute  # noqa
from util.team.team import teamid_to_encoding  # noqa

_divisions = divisions()
_fairylab_root = re.sub(r'/filefairy', '/fairylab/static', _root)
_game_path = '/resource/games/game_{}.json'
_html = 'https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/'
_player = 'players/player_{}.html'
_player_default = {
    'name': 'Jim Unknown',
    'number': '0',
    'bats': '-',
    'throws': '-'
}
_smallcaps = {'L': 'ʟ', 'R': 'ʀ', 'S': 'ꜱ'}


class Gameday(Registrable):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.players = {}

    @staticmethod
    def _data():
        return os.path.join(_path, 'data.json')

    @staticmethod
    def _href():
        return '/gameday/'

    @staticmethod
    def _info():
        return 'Replays finished games in real time.'

    @staticmethod
    def _title():
        return 'gameday'

    def _notify_internal(self, **kwargs):
        if kwargs['notify'] == Notify.STATSPLUS_SIM:
            self.data['started'] = False
            self.data['games'] = []
            self.write()
        return Response()

    def _on_message_internal(self, **kwargs):
        return Response()

    def _render_internal(self, **kwargs):
        recreate(_fairylab_root + '/gameday/')

        games = copy.deepcopy(self.data['games'])
        schedule_data = self._schedule_data(games)

        ret = []

        gameday = self._gameday(schedule_data)
        html = 'gameday/index.html'
        ret.append((html, '', 'gameday.html', gameday))

        for id_ in games:
            with open(_root + _game_path.format(id_), 'r') as f:
                game_data = json.loads(f.read())
                away_team = encoding_to_nickname(game_data['away_team'])
                home_team = encoding_to_nickname(game_data['home_team'])
                date = decode_datetime(game_data['date']).strftime('%m/%d/%Y')
                subtitle = '{} at {}, {}'.format(away_team, home_team, date)
                game = self._game(id_, subtitle, game_data, schedule_data)
                html = 'gameday/{}/index.html'.format(id_)
                ret.append((html, subtitle, 'game.html', game))

        return ret

    def _run_internal(self, **kwargs):
        response = Response()
        if self._check_games():
            self._render(**kwargs)
            response.append_notify(Notify.BASE)
        return response

    def _setup_internal(self, **kwargs):
        self._render(**kwargs)
        return Response()

    def _shadow_internal(self, **kwargs):
        return []

    def _game_repl(self, game_data, m):
        a = m.group(0)
        if a.startswith('P'):
            return self.players[a]['name'] if a in self.players else a
        if a.startswith('T'):
            return encoding_to_decoding(a)
        return a

    def _game_sub(self, game_data):
        pattern = '|'.join([id_ for id_ in self.players] +
                           [game_data['away_team'], game_data['home_team']])
        return partial(re.sub, pattern, partial(self._game_repl, game_data))

    @staticmethod
    def _schedule_data(games):
        schedule_data = {}
        for id_ in games:
            with open(_root + _game_path.format(id_), 'r') as f:
                game_data = json.loads(f.read())
                date = decode_datetime(game_data['date'])

                away_team = game_data['away_team']
                if away_team not in schedule_data:
                    schedule_data[away_team] = []

                home_team = game_data['home_team']
                if home_team not in schedule_data:
                    schedule_data[home_team] = []

                schedule_data[away_team].append((date, home_team, '@', id_))
                schedule_data[home_team].append((date, away_team, 'v', id_))

        for encoding in schedule_data:
            schedule_data[encoding] = sorted(schedule_data[encoding])

        return schedule_data

    @staticmethod
    def _schedule_head(decoding):
        return table(
            clazz='table-fixed border border-bottom-0 mt-3', head=[decoding])

    @staticmethod
    def _schedule_body(encoding, id_, schedule_data):
        body = []
        for sdate, steam, ssymbol, sid in schedule_data[encoding]:
            sdate = sdate.strftime('%m/%d/%Y')
            steam = encoding_to_decoding(steam)
            stext = '{} {} {}'.format(sdate, ssymbol, steam)
            if id_ == sid:
                body.append([secondary(stext)])
            else:
                url = '/gameday/{}/'.format(sid)
                body.append([anchor(url, stext)])

        return table(clazz='table-fixed border', body=body)

    def _add_players(self, players):
        for id_ in players:
            if 'P' + id_ in self.players:
                continue
            link = _html + _player.format(id_)
            player = parse_player(link)
            self.players['P' + id_] = player

    def _check_games(self):
        games = []
        for game in os.listdir(_root + '/resource/games/'):
            id_ = re.findall('game_(\d+).json', game)[0]
            games.append(id_)

        games = sorted(games)
        if games != self.data['games']:
            self.data['games'] = games

            if not self.data['started']:
                self.data['started'] = True
                self._chat('fairylab', 'Live sim created.')

            self.write()
            return True

        return False

    def _gameday(self, schedule_data):
        ret = {
            'breadcrumbs': [{
                'href': '/',
                'name': 'Fairylab'
            }, {
                'href': '',
                'name': 'Gameday'
            }],
            'schedule': []
        }

        for division, ts in _divisions:
            division = division.replace('AL', 'American League')
            division = division.replace('NL', 'National League')
            ret['schedule'].append(
                table(
                    clazz='table-fixed border border-bottom-0 mt-3',
                    head=[division]))
            body = []
            for teamid in ts:
                encoding = teamid_to_encoding(teamid)
                decoding = encoding_to_decoding(encoding)
                if encoding in schedule_data:
                    sid = schedule_data[encoding][0][3]
                    stext = anchor('/gameday/{}/'.format(sid), decoding)
                else:
                    stext = secondary(decoding)
                body.append([logo_absolute(teamid, stext, 'left')])
            ret['schedule'].append(
                table(
                    clazz='table-fixed border',
                    bcols=[' class="position-relative text-truncate"'],
                    body=body))

        return ret

    @staticmethod
    def _badge(pitch, sequence):
        p = '<div class="badge badge-pill alert-pitch alert-{0}">{1}</div>'
        if 'In play' in sequence:
            return p.format('primary', pitch) + sequence
        if 'Ball' in sequence:
            return p.format('success', pitch) + sequence
        return p.format('danger', pitch) + sequence

    @staticmethod
    def _profile(name, s):
        ins = ''.join([word[0].upper() for word in name.split(' ', 2)])
        div = '<div class="profile position-absolute">{}</div>'.format(ins)
        span = '<span class="align-middle d-block pl-60p">{}</span>'.format(s)
        return div + span

    def _atbat(self, id_):
        if id_ not in self.players:
            player = _player_default
        else:
            player = self.players[id_]
        name = player['name']
        s = 'ᴀᴛ ʙᴀᴛ: #{} ({})<br>{}'.format(
            player['number'], _smallcaps.get(player['bats'], 'ʀ'), name)
        return self._profile(name, s)

    def _pitching(self, id_):
        if id_ not in self.players:
            player = _player_default
        else:
            player = self.players[id_]
        name = player['name']
        s = 'ᴘɪᴛᴄʜɪɴɢ: #{} {}ʜᴘ<br>{}'.format(
            player['number'], _smallcaps.get(player['throws'], 'ʀ'), name)
        return self._profile(name, s)

    def _game(self, game_id_, subtitle, game_data, schedule_data):
        ret = {
            'breadcrumbs': [{
                'href': '/',
                'name': 'Fairylab'
            }, {
                'href': '/gameday/',
                'name': 'Gameday'
            }, {
                'href': '',
                'name': subtitle
            }],
            'tabs': {
                'style': 'tabs',
                'tabs': []
            }
        }

        self._add_players(game_data['players'])

        game_sub = self._game_sub(game_data)
        away_team = game_data['away_team']
        away_decoding = encoding_to_decoding(away_team)
        home_team = game_data['home_team']
        home_decoding = encoding_to_decoding(home_team)

        log_tables = []
        plays = {
            'name': 'plays',
            'title': 'Plays',
            'tabs': {
                'style': 'pills',
                'tabs': []
            }
        }
        for i, inning in enumerate(game_data['plays']):
            plays_tables = []
            for half in inning:
                teamid = encoding_to_teamid(half['batting'])
                log_table = table(
                    hcols=[' colspan="2" class="position-relative"'],
                    head=[logo_absolute(teamid, half['label'], 'left')],
                    bcols=[
                        ' class="position-relative"',
                        ' class="text-center text-secondary w-55p"'
                    ],
                    body=[])
                plays_table = table(
                    hcols=[' class="position-relative"'],
                    head=[logo_absolute(teamid, half['label'], 'left')],
                    body=[])
                outs = 0
                for play in half['play']:
                    if play['type'] == 'sub':
                        value = game_sub(play['value'])
                        if play['subtype'] == 'pitching':
                            log_table['body'].append(
                                [self._pitching(play['value']), ''])
                            plays_table['body'].append(['Pitching: ' + value])
                        elif play['subtype'] == 'batting':
                            log_table['body'].append(
                                [self._atbat(play['value']), ''])
                        else:
                            log_table['body'].append([value, ''])
                    elif play['type'] == 'event':
                        for s in play['sequence']:
                            pitch, balls, strikes, value = s.split(' ', 3)
                            badge = self._badge(pitch, value)
                            count = '' if 'In play' in s else '{}-{}'.format(
                                balls, strikes)
                            log_table['body'].append([badge, count])
                        value = game_sub(play['value'])
                        if play['outs']:
                            outs += play['outs']
                            value += ' ' + bold('{} out.'.format(outs))
                        log_table['body'].append([value, ''])
                        if outs < 3:
                            log_table['body'].append(['&nbsp;', '&nbsp;'])
                        plays_table['body'].append([value])
                if half['footer']:
                    log_table['fcols'] = [' colspan="2"']
                    log_table['foot'] = [game_sub(half['footer'])]
                    plays_table['foot'] = [game_sub(half['footer'])]
                log_tables.append(log_table)
                plays_tables.append(plays_table)
            plays['tabs']['tabs'].append({
                'name': 'plays-' + str(i + 1),
                'title': str(i + 1),
                'tables': plays_tables
            })
        ret['tabs']['tabs'].append({
            'name': 'log',
            'title': 'Game Log',
            'tables': log_tables
        })

        schedule_tables = []
        schedule_tables.append(self._schedule_head(away_decoding))
        schedule_tables.append(
            self._schedule_body(away_team, game_id_, schedule_data))
        schedule_tables.append(self._schedule_head(home_decoding))
        schedule_tables.append(
            self._schedule_body(home_team, game_id_, schedule_data))
        ret['tabs']['tabs'].append({
            'name': 'schedule',
            'title': 'Schedule',
            'tables': schedule_tables
        })

        ret['tabs']['tabs'].append(plays)
        return ret


# from plugin.statsplus.statsplus import Statsplus
# from util.datetime_.datetime_ import datetime_now
# from util.jinja2_.jinja2_ import env

# date = datetime_now()
# e = env()
# statsplus = Statsplus(date=date, e=e)

# for encoded_date in statsplus.data['scores']:
#     for score in statsplus.data['scores'][encoded_date]:
#         id_ = re.findall('(\d+)\.html', score)[0]
#         statsplus._extract(encoded_date, id_)
# statsplus._extract('2024-06-02T00:00:00-07:00', '1298')

# gameday = Gameday(date=date, e=e)
# gameday.data['games'] = ['1298']
# gameday._check_games()
# gameday._setup_internal(date=date)
