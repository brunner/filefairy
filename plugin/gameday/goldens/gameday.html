<!doctype html>
<html lang="en">

<head>

<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link rel="stylesheet"
      href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css"
      integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm"
      crossorigin="anonymous">
<link rel="stylesheet"
      href="https://cdnjs.cloudflare.com/ajax/libs/open-iconic/1.1.1/font/css/open-iconic-bootstrap.min.css">
<title>Fairylab | gameday</title>

<style>
html {
  position: relative;
  min-height: 100%;
}
body {
  margin-bottom: 60px;
}
header {
  background-color: #f5f5f5;
  width: 100%;
}
a.disabled{
  pointer-events: none;
  cursor: default;
}
.alert, .dropdown-toggle, .form-control {
  padding: .75rem;
}
.alert .oi {
  line-height: 24px;
  padding-right: .75rem;
}
.anchor {
  padding-top: 0px;
  margin-top: 0px;
  transition: padding-top .35s ease, margin-top .35s ease;
}
.breadcrumb {
  background-color: inherit;
  line-height: 60px;
  max-width: 468px;
  overflow-x: auto;
}
.dropdown-toggle {
  color: #6c767c;
}
.flex-grow {
  flex-grow: 1;
}
.left-8p {
  left: 8px;
}
.navbar {
  padding: 0;
}
.pl-24p {
  padding-left: 24px;
}
.pr-24p {
  padding-right: 24px;
}
.right-8p {
  right: 8px;
}
.table, .table-sm {
  margin-bottom: 0;
  word-break: break-word;
}
.table .thead-light td {
  background-color: #f5f5f5;
}
.table-fixed {
  table-layout: fixed;
}
.table-sm td {
  border-top: 0;
  padding: 0 .3rem 0 0;
}
.table .td-sm {
  padding: .75rem .25rem;
}
.table.collapse.show {
  display: table;
}
.top-14p {
  top: 14px;
}
.w-10 {
  width: 10%;
}
.w-20 {
  width: 20%;
}
.w-40 {
  width: 40%;
}
.w-55p {
  width: 55px;
}
.w-65p {
  width: 65px;
}
.w-75p {
  width: 75px;
}
.w-105p {
  width: 105px;
}
</style>
</head>

<body>

<!-- ^(base.header) -->
<header class="border-bottom sticky-top">
<div class="container">
<nav class="navbar navbar-light border-bottom" aria-label="nav">
<ol class="breadcrumb flex-nowrap text-truncate m-0 p-0">
<li class="breadcrumb-item">
<a href="/">Fairylab</a>
</li>
<li class="breadcrumb-item active" aria-current="page">Gameday</li>
</ol>
</nav>

<div class="my-1">
<small class="text-muted">Last reloaded: 06:02:30 PDT (1985-10-26)</small>
</div>

</div>
</header>
<!-- $(base.header) -->

<!-- ^(base.main) -->
<main role="main" class="pb-3">

<!-- ^(gameday.schedule) -->
<div class="container anchor" id="schedule">

<!-- ^(common.heading) -->
<div class="row">
<div class="col-12">
<div class="alert alert-secondary mt-3 mb-0">
<span class="oi oi-calendar text-secondary"></span>Schedule
</div>
</div>
</div>
<!-- $(common.heading) -->

<!-- ^(gameday.row) -->
<div class="row">

<!-- ^(gameday.column) -->
<div class="col-12">

<!-- ^(common.table) -->
<table class="table table-fixed border border-bottom-0 mt-3">

<!-- ^(common.table.head) -->
<thead class="thead-light">
<tr>
<td scope="col">American League East</td>
</tr>
</thead>
<!-- $(common.table.head) -->

</table>
<!-- $(common.table) -->

<!-- ^(common.table) -->
<table class="table table-fixed border">

<!-- ^(common.table.body) -->
<tbody>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/baltimore_orioles_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/1/">Baltimore Orioles</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/boston_red_sox_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/2/">Boston Red Sox</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/new_york_yankees_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/3/">New York Yankees</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/tampa_bay_rays_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/4/">Tampa Bay Rays</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/toronto_blue_jays_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/5/">Toronto Blue Jays</a></span></td>
</tr>
</tbody>
<!-- $(common.table.body) -->

</table>
<!-- $(common.table) -->

<!-- ^(common.table) -->
<table class="table table-fixed border border-bottom-0 mt-3">

<!-- ^(common.table.head) -->
<thead class="thead-light">
<tr>
<td scope="col">American League Central</td>
</tr>
</thead>
<!-- $(common.table.head) -->

</table>
<!-- $(common.table) -->

<!-- ^(common.table) -->
<table class="table table-fixed border">

<!-- ^(common.table.body) -->
<tbody>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/chicago_white_sox_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/1/">Chicago White Sox</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/cleveland_indians_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/2/">Cleveland Indians</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/detroit_tigers_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/3/">Detroit Tigers</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/kansas_city_royals_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/4/">Kansas City Royals</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/minnesota_twins_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/5/">Minnesota Twins</a></span></td>
</tr>
</tbody>
<!-- $(common.table.body) -->

</table>
<!-- $(common.table) -->

<!-- ^(common.table) -->
<table class="table table-fixed border border-bottom-0 mt-3">

<!-- ^(common.table.head) -->
<thead class="thead-light">
<tr>
<td scope="col">American League West</td>
</tr>
</thead>
<!-- $(common.table.head) -->

</table>
<!-- $(common.table) -->

<!-- ^(common.table) -->
<table class="table table-fixed border">

<!-- ^(common.table.body) -->
<tbody>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/houston_astros_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/1/">Houston Astros</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/los_angeles_angels_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/2/">Los Angeles Angels</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/oakland_athletics_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/3/">Oakland Athletics</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/seattle_mariners_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/4/">Seattle Mariners</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/texas_rangers_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/5/">Texas Rangers</a></span></td>
</tr>
</tbody>
<!-- $(common.table.body) -->

</table>
<!-- $(common.table) -->

<!-- ^(common.table) -->
<table class="table table-fixed border border-bottom-0 mt-3">

<!-- ^(common.table.head) -->
<thead class="thead-light">
<tr>
<td scope="col">National League East</td>
</tr>
</thead>
<!-- $(common.table.head) -->

</table>
<!-- $(common.table) -->

<!-- ^(common.table) -->
<table class="table table-fixed border">

<!-- ^(common.table.body) -->
<tbody>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/atlanta_braves_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/1/">Atlanta Braves</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/miami_marlins_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/2/">Miami Marlins</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/new_york_mets_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/3/">New York Mets</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/philadelphia_phillies_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/4/">Philadelphia Phill</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/washington_nationals_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/5/">Washington Nationa</a></span></td>
</tr>
</tbody>
<!-- $(common.table.body) -->

</table>
<!-- $(common.table) -->

<!-- ^(common.table) -->
<table class="table table-fixed border border-bottom-0 mt-3">

<!-- ^(common.table.head) -->
<thead class="thead-light">
<tr>
<td scope="col">National League Central</td>
</tr>
</thead>
<!-- $(common.table.head) -->

</table>
<!-- $(common.table) -->

<!-- ^(common.table) -->
<table class="table table-fixed border">

<!-- ^(common.table.body) -->
<tbody>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/chicago_cubs_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/1/">Chicago Cubs</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/cincinnati_reds_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/2/">Cincinnati Reds</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/milwaukee_brewers_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/3/">Milwaukee Brewers</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/pittsburgh_pirates_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/4/">Pittsburgh Pirates</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/st_louis_cardinals_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/5/">St. Louis Cardinal</a></span></td>
</tr>
</tbody>
<!-- $(common.table.body) -->

</table>
<!-- $(common.table) -->

<!-- ^(common.table) -->
<table class="table table-fixed border border-bottom-0 mt-3">

<!-- ^(common.table.head) -->
<thead class="thead-light">
<tr>
<td scope="col">National League West</td>
</tr>
</thead>
<!-- $(common.table.head) -->

</table>
<!-- $(common.table) -->

<!-- ^(common.table) -->
<table class="table table-fixed border">

<!-- ^(common.table.body) -->
<tbody>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/arizona_diamondbacks_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/1/">Arizona Diamondbac</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/colorado_rockies_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/2/">Colorado Rockies</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/los_angeles_dodgers_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/3/">Los Angeles Dodger</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/san_diego_padres_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/4/">San Diego Padres</a></span></td>
</tr>
<tr>
<td scope="row" class="position-relative text-truncate"><img src="https://orangeandblueleaguebaseball.com/StatsLab/reports/news/html/images/team_logos/san_francisco_giants_40.png" width="20" height="20" border="0" class="position-absolute left-8p top-14p"><span class="align-middle d-block text-truncate pl-24p"><a href="/gameday/5/">San Francisco Gian</a></span></td>
</tr>
</tbody>
<!-- $(common.table.body) -->

</table>
<!-- $(common.table) -->

</div>
<!-- $(gameday.column) -->

</div>
<!-- $(gameday.row) -->

</div>
<!-- $(gameday.schedule) -->

</main>
<!-- $(base.main) -->

<script src="https://code.jquery.com/jquery-3.2.1.slim.min.js"
        integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN"
        crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js"
        integrity="sha384-ApNbgh9B+Y1QKtv3Rn7W3mgPxhU9K/ScQsAP7hUibX39j7fakFPskvXusvfa0b4Q"
        crossorigin="anonymous"></script>
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta.3/js/bootstrap.min.js"
        integrity="sha384-a5N7Y/aK3qNeh15eJKGWxsqtnX/wWdSZSKp+81YjTmS15nvnvxKHuzaWwXHDli+4"
        crossorigin="anonymous"></script>
<script>
$('.repl-toggler').on('click', function(e) {
  if ($(this).hasClass('disabled')) {
    e.preventDefault();
  } else {
    $(this).toggleClass('disabled text-secondary');
    $(this).html($(this).data('repl'));
  }
});
$('.show-toggler').on('click', function(e) {
  $(this).toggleClass('border-bottom-0');
});
</script>


</body>

</html>